From 6c018663fe706dffb8b76b23831433ae3e55dbe0 Mon Sep 17 00:00:00 2001
From: Daniel Emminizer <daniel.m.emminizer.civ@us.navy.mil>
Date: Thu, 20 Mar 2025 11:27:30 +0000
Subject: [PATCH 13/29] SDK: Install Protobuf Shared Objects (1/2)

**JIRA Issue:** SIM-15139

**Description:** If Protobuf 29.3 shared objects / DLLs are found, install them to lib. Fixed ZLib directory name. Fixed issue with standalone Linux build where ImGui depends on Threads (pthreads) but protobuf as a shared library may not transitively carry it forward.

**Testing Performed:** Make install and ldd of protobuf-related libraries.
---
 CMakeLists.txt                                |  1 +
 CMakeUtilities/InstallData.3rd.Protobuf.cmake | 35 +++++++++++++++++++
 CMakeUtilities/VsiPackages.json               |  1 +
 Examples/imgui/CMakeLists.txt                 |  3 ++
 4 files changed, 40 insertions(+)
 create mode 100644 CMakeUtilities/InstallData.3rd.Protobuf.cmake

diff --git a/CMakeLists.txt b/CMakeLists.txt
index b823f37a..3dc447de 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -240,6 +240,7 @@ if(INSTALL_THIRDPARTY_LIBRARIES)
     include(InstallData.3rd.OSG)
     include(InstallData.3rd.osgEarth)
     include(InstallData.3rd.osgQt)
+    include(InstallData.3rd.Protobuf)
     include(InstallData.3rd.Qt5)
 endif()
 
diff --git a/CMakeUtilities/InstallData.3rd.Protobuf.cmake b/CMakeUtilities/InstallData.3rd.Protobuf.cmake
new file mode 100644
index 00000000..34b6d39d
--- /dev/null
+++ b/CMakeUtilities/InstallData.3rd.Protobuf.cmake
@@ -0,0 +1,35 @@
+if(NOT Protobuf_FOUND)
+    find_package(Protobuf CONFIG QUIET)
+    if(NOT Protobuf_FOUND)
+        find_package(Protobuf QUIET)
+    endif()
+endif()
+if(NOT Protobuf_FOUND)
+    return()
+endif()
+
+# Try to get Protobuf directory
+get_target_property(_PROTO_SO protobuf::libprotobuf IMPORTED_LOCATION_RELEASE)
+if(NOT _PROTO_SO)
+    get_target_property(_PROTO_SO protobuf::libprotobuf IMPORTED_LOCATION)
+endif()
+get_filename_component(_PROTO_LIB_DIR "${_PROTO_SO}" DIRECTORY)
+get_filename_component(_PROTO_DIR "${_PROTO_LIB_DIR}" DIRECTORY)
+
+if(WIN32)
+    vsi_install_target(GDAL::GDAL ThirdPartyLibs)
+    set(_PROTO_BIN ${_PROTO_DIR}/bin)
+    foreach(_LIB IN ITEMS abseil_dll libprotobuf utf8_range utf8_validity)
+        install(PROGRAMS ${_PROTO_BIN}/${_LIB}.dll DESTINATION "${INSTALLSETTINGS_SHARED_LIBRARY_DIR}"
+            COMPONENT ThirdPartyLibs CONFIGURATIONS Release RelWithDebInfo MinSizeRel "" OPTIONAL)
+        install(PROGRAMS ${_PROTO_BIN}/${_LIB}d.dll DESTINATION "${INSTALLSETTINGS_SHARED_LIBRARY_DIR}"
+            COMPONENT ThirdPartyLibs CONFIGURATIONS DEBUG OPTIONAL)
+    endforeach()
+else()
+    foreach(_LIB IN ITEMS abseil_dll protobuf utf8_range utf8_validity)
+        if(EXISTS "${_PROTO_LIB_DIR}/lib${_LIB}.so")
+            get_symlinks("${_PROTO_LIB_DIR}/lib${_LIB}.so" _LIBS)
+            install(PROGRAMS ${_LIBS} DESTINATION "${INSTALLSETTINGS_SHARED_LIBRARY_DIR}" COMPONENT ThirdPartyLibs)
+        endif()
+    endforeach()
+endif()
diff --git a/CMakeUtilities/VsiPackages.json b/CMakeUtilities/VsiPackages.json
index 90b38332..efc7576f 100644
--- a/CMakeUtilities/VsiPackages.json
+++ b/CMakeUtilities/VsiPackages.json
@@ -78,6 +78,7 @@
     },
     {
       "name": "ZLIB",
+      "dir": "ZLib",
       "version": [
         "1.3.1",
         "1.2.13"
diff --git a/Examples/imgui/CMakeLists.txt b/Examples/imgui/CMakeLists.txt
index bef035ee..0adc36c4 100644
--- a/Examples/imgui/CMakeLists.txt
+++ b/Examples/imgui/CMakeLists.txt
@@ -64,3 +64,6 @@ set_target_properties(ImGuiLib PROPERTIES
     FOLDER "ImGui Library"
     PROJECT_LABEL "ImGui"
 )
+
+find_package(Threads)
+target_link_libraries(ImGuiLib PRIVATE Threads::Threads)
-- 
2.47.1.windows.2

